File: 0039_Combination Sum/CombinationSum.cs
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        
        var ans = new List<IList<int>>();
        if(candidates == null || candidates.Length == 0) return ans;
        
        Array.Sort(candidates);
        DFS(candidates, 0, target, new List<int>(), ans);
        
        return ans;
    }
    
    private void DFS(int[] candidates, int s, int target, IList<int> list, IList<IList<int>> ans){
        
            if(target == 0){
                ans.Add(new List<int>(list));
                return;
            } 
                
        
        for(int i=s;i<candidates.Length;i++){
            if(candidates[i] > target) break;

            list.Add(candidates[i]);   
            DFS(candidates, i,target - candidates[i], list, ans);
            list.RemoveAt(list.Count - 1);
        }
    }
}

File: 0039_Combination Sum/CombinationSum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();

        List<Integer> cur = new ArrayList<Integer>();
        Arrays.sort(candidates);
        DFS(candidates, target, 0, cur, ans);
        return ans;
    }

    private void DFS(int[] candidates, int target, int s, List<Integer> cur, List<List<Integer>> ans){
        if(target == 0){
            ans.add(new ArrayList<Integer>(cur));
            return;
        }

        for(int i=s;i<candidates.length;i++){
            if(candidates[i] > target) break;
            if (i > s && candidates[i] == candidates[i-1]) continue;
            cur.add(candidates[i]);
            DFS(candidates, target - candidates[i], i, cur, ans);
            cur.remove(cur.size()-1);
        }
    }
}